### PR TITLE
Fixed atmospherics components air/unsafe pressure releases

### PR DESCRIPTION
**Pull request by mogeoko**
## About The Pull Request

At present

- If user deconstructs/unwrenches something from [/obj/machinery/atmospherics/components](https://github.com/tgstation/tgstation/blob/master/code/modules/atmospherics/machinery/components/components_base.dm) - all air inside it disappear. (only unary, binary, trinary devices).

- [/obj/machinery/atmospherics/components/unsafe_pressure_release()](https://github.com/tgstation/tgstation/blob/f9a5502f78263690378ca01f660b1cfb2886efd1/code/modules/atmospherics/machinery/components/components_base.dm#L175) doesn't work at all.

- HFR UI has the problem of [activation request message](https://github.com/tgstation/tgstation/blob/f9a5502f78263690378ca01f660b1cfb2886efd1/code/modules/atmospherics/machinery/components/fusion/hfr_parts.dm#L209C3-L209C3) constantly repeating if the user forcefully switches off the machine (e.g. unscrewed one of the components).

- Thermomachine and Bluespace sender can't be connected to pipes when panel is open.

Now 

- Each [/obj/machinery/atmospherics/components](https://github.com/tgstation/tgstation/blob/master/code/modules/atmospherics/machinery/components/components_base.dm) deconstruction will move air into connected pipeline if any, otherwise it will be released from open nodes, HFR internals or if there is portable connector that has no device connected. The user will also be notified of deconstruction and if pressure inside is safe for action.

- The [/obj/machinery/atmospherics/components/unsafe_pressure_release()](https://github.com/tgstation/tgstation/blob/f9a5502f78263690378ca01f660b1cfb2886efd1/code/modules/atmospherics/machinery/components/components_base.dm#L175) will act like a normal pipe [unsafe_pressure_release()](https://github.com/tgstation/tgstation/blob/f9a5502f78263690378ca01f660b1cfb2886efd1/code/modules/atmospherics/machinery/atmosmachinery.dm#L447C1-L447C1).

- When HFR interface is turned off - UI will be closed.

- Thermomachine and Bluespace sender will always try to connect to pipe if they are anchored

## Why It's Good For The Game

Removing all air inside components is not very nice. This also closes the problem with infinite gas removal. 

## Changelog

:cl: mogeoko
fix: Atmospherics components will now move air into connected pipeline on deconstruction if possible. Otherwise, air will be released to the outside from open nodesrelease air outside.
fix: Unsafe pressure release on atmos components will now work the same way it does in the normal pipes.
fix: The HFR user interface would close when the machine is shut down.
/:cl:

### CODE REVIEW

File: code/modules/atmospherics/machinery/components/components_base.dm [hunk 1]
```diff
@@ -32,6 +32,24 @@
...
+				parents_air.merge(air)
+				continue
+		if(!to_release)
+			to_release = air
+			continue
+		to_release.merge(air)
+	if(to_release)
+		T.assume_air(to_release)
+	return ..()
+
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/components/components_base.dm:52]: Destroy should not be performing any side effects, it should delete the object and nothing more.

What are you trying to accomplish with this?

File: code/modules/atmospherics/machinery/components/fusion/hfr_core.dm [hunk 2]
```diff
@@ -181,9 +181,12 @@
 
 /obj/machinery/atmospherics/components/unary/hypertorus/core/Destroy()
 	unregister_signals(TRUE)
+	var/turf/T = loc
 	if(internal_fusion)
+		T.assume_air(internal_fusion)
 		internal_fusion = null
 	if(moderator_internal)
+		T.assume_air(moderator_internal)
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/components/fusion/hfr_core.dm:189]: Same

File: code/modules/atmospherics/machinery/components/fusion/hfr_parts.dm [hunk 3]
```diff
@@ -40,10 +40,49 @@
...
 
+/obj/machinery/atmospherics/components/unary/hypertorus/crowbar_act(mob/user, obj/item/tool)
+	if(!panel_open)
+		balloon_alert(user, "open panel!")
+		return TOOL_ACT_TOOLTYPE_SUCCESS
+
+	var/unsafe_wrenching = FALSE
+	var/filled_pipe = FALSE
+	var/datum/gas_mixture/env_air = loc.return_air()
+	var/datum/gas_mixture/ins_air = airs[1]
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/components/fusion/hfr_parts.dm:53]: Do not abbreviate names. I have no idea what "ins" means.

File: code/modules/atmospherics/machinery/components/fusion/hfr_parts.dm [hunk 4]
```diff
@@ -40,10 +40,49 @@
...
+		internal_pressure = max(internal_pressure, ins_air.return_pressure())	
+	
+	internal_pressure -= env_air.return_pressure()
+
+	if(ins_air.total_moles() > 0)
+		filled_pipe = TRUE
+
+	default_deconstruction_crowbar(tool, custom_deconstruct = filled_pipe)
+		
+	to_chat(user, span_notice("You begin to unfasten \the [src]..."))
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/components/fusion/hfr_parts.dm:71]:
```suggestion
	balloon_alert(user, "unfastening...")
```

File: code/modules/atmospherics/machinery/components/fusion/hfr_parts.dm [hunk 5]
```diff
@@ -40,10 +40,49 @@
...
+	
+	if(internal_pressure > 2 * ONE_ATMOSPHERE)
+		to_chat(user, span_warning("As you begin deconstructing \the [src] a gush of air blows in your face... maybe you should reconsider?"))
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/components/fusion/hfr_parts.dm:74]: This should be a balloon alert, though I know the others aren't. This should either be in the same balloon alert as "unfastening...", or sent after a very short delay.

File: code/modules/atmospherics/machinery/components/fusion/hfr_parts.dm [hunk 6]
```diff
@@ -40,10 +40,49 @@
...
+		unsafe_wrenching = TRUE
+	
+	if(!do_after(user, 2 SECONDS, src))
+		return
+	if(unsafe_wrenching)
+		unsafe_pressure_release(user, internal_pressure)
+	
+	tool.play_tool_sound(src, 50)
+	deconstruct(TRUE)
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/components/fusion/hfr_parts.dm:83]: Named parameter for TRUE

File: code/modules/atmospherics/machinery/components/gas_recipe_machines/crystallizer.dm [hunk 7]
```diff
@@ -59,10 +59,42 @@
...
+			var/pressure_delta = airs[i].return_pressure() - env_air.return_pressure()
+			internal_pressure = internal_pressure > pressure_delta ? internal_pressure : pressure_delta
+
+	default_deconstruction_crowbar(tool, custom_deconstruct = filled_pipe)
+	
+	to_chat(user, span_notice("You begin to unfasten \the [src]..."))
+
+	if(internal_pressure > 2 * ONE_ATMOSPHERE)
+		to_chat(user, span_warning("As you begin deconstructing \the [src] a gush of air blows in your face... maybe you should reconsider?"))
+		unsafe_wrenching = TRUE
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/components/gas_recipe_machines/crystallizer.dm:87]: Same

File: code/modules/atmospherics/machinery/components/gas_recipe_machines/crystallizer.dm [hunk 8]
```diff
@@ -59,10 +59,42 @@
...
+		
+	if(!do_after(user, 2 SECONDS, src))
+		return
+	if(unsafe_wrenching)
+		unsafe_pressure_release(user, internal_pressure)
+	
+	tool.play_tool_sound(src, 50)
+	deconstruct(TRUE)
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/components/gas_recipe_machines/crystallizer.dm:95]: Same

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 9]
```diff
@@ -457,9 +457,36 @@
...
+		to_chat(user, span_warning("As you begin deconstructing \the [src] a gush of air blows in your face... maybe you should reconsider?"))
+		unsafe_wrenching = TRUE
+	
+	if(!do_after(user, 2 SECONDS, src))
+		return
+	if(unsafe_wrenching)
+		internal_connector.gas_connector.unsafe_pressure_release(user, internal_pressure)
+	
+	tool.play_tool_sound(src, 50)
+	deconstruct(TRUE)
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:487]: Same, but also really don't like how this is all copied and pasted

File: code/modules/atmospherics/machinery/components/unary_devices/thermomachine.dm [hunk 10]
```diff
@@ -215,7 +215,35 @@
...
+		to_chat(user, span_warning("As you begin deconstructing \the [src] a gush of air blows in your face... maybe you should reconsider?"))
+		unsafe_wrenching = TRUE
+	
+	if(!do_after(user, 2 SECONDS, src))
+		return
+	if(unsafe_wrenching)
+		unsafe_pressure_release(user, internal_pressure)
+	
+	tool.play_tool_sound(src, 50)
+	deconstruct(TRUE)
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/components/unary_devices/thermomachine.dm:245]: Too much copy and paste

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by Mothblocks** [comment]
Can you also improve your title to make it more obvious what this PR actually is doing?

**Comment by GPeckman** [comment]
I honestly thought deleting the gas was a feature.

**Comment by MTandi** [comment]
You need to do something with thermomachines then. I assume that after this PR you can no longer have a safe setup of thermomachine and a port. It will always have some leftover gas, and there is no way to extract it other than pumping into a small tank with a portable atmos pump.

**Comment by FlufflesTheDog** [comment]
Been wanting to do something like this for a long time. Always bothered me that devices could just delete gas, and having it try to go into the pipenet first is great for still allowing maintainence without making a huge mess.

Does this actually modify things like cryo or thermomachines as well, which currently delete their air contents when deconstructed?

**Comment by mogeoko** [comment]
> Been wanting to do something like this for a long time. Always bothered me that devices could just delete gas, and having it try to go into the pipenet first is great for still allowing maintainence without making a huge mess.
> 
> Does this actually modify things like cryo or thermomachines as well, which currently delete their air contents when deconstructed?

Cryo not yet, but thermomachines (not for now), crystallizer and HFR - yes.

**Comment by SyncIt21** [comment]
> deconstruction will move air into connected pipeline

This should not be allowed, the only components that should ever be allowed to "move air" are pumps and not a wrench.

Atmos components releasing air on the turf when deconstructed is good but does it fix?
- #47556

> This also closes the problem with infinite gas removal.

Can you elaborate more on what the problem actually is? Are you talking about?
- #73042 

Better call in the atmos expert. @LemonInTheDark

File: code/modules/atmospherics/machinery/components/components_base.dm (hunk 1 above)
Comment (SyncIt21) [reply code/modules/atmospherics/machinery/components/components_base.dm:52]: I agree. Override  and move all this code to on_deconstruct(). 

So this code is responsible for merging/moving air from this pipe to its connected neighbour's when it is destroyed but again as @MTandi has pointed out in https://github.com/tgstation/tgstation/pull/79924#issuecomment-1826164364 this causes problems

**Comment by Deadgebert** [comment]
I like the idea of it pushing gas into connected networks, but i'm heavily opposed to them releasing their gasses. it doesn't really add much to the game and is more tedium added for no real reason other than 'it makes sense physically'

**Comment by mogeoko** [comment]
Thank you for your comments, I have corrected the code you pointed out.

First I would like to address the topic of moving air. On the one hand, it shouldn't be moved between nets when component is disassembled, but Deadgebert's comment correctly points out that there's no actual reason why air has to be thrown outside (other than a simulation of physics) and it will only cause harm to players by forcing them to do unnecessary actions in the name of realism.
With my fix, players will only feel the negative effects of their own mistakes while saving their valuable work.

>Atmos components releasing air on the turf when deconstructed is good but does it fix?

No, that has nothing to do with this problem, my code only affects components.

>Can you elaborate more on what the problem actually is? Are you talking about?

No, right now anyone can dedicate a couple minutes of their game to simply removing a huge portion of gases without any consequences. An example of this is the constant installing and disassembling of machinery into pipes.

**Comment by mogeoko** [comment]
> I don't see why intentionally deleting gas can really be considered an issue when it can be vented into space or ice turf instead. If I wanted to delete the gas in atmos, I'd grab the fireaxe and smash into the gas tanks.

The same can be said for the release of air when normal pipes are disassembled. It is quite unpleasant to lose a part of some gas at each interaction with components and this can be used against others.

> I looked through the changes and I may have missed it, but how does this handle connection ports being directly connected to an atmos device?

Components connected to connectors will only release air if there are no devices on the connector itself. If there are, they will go inside.

[Example](https://github.com/tgstation/tgstation/blob/ebd3db3e3197bef8373dd144f9513dd4e0357611/code/modules/atmospherics/machinery/components/components_base.dm#L341)

**Comment by Deadgebert** [comment]
i don't think these are bugs and should be considered balances instead of fixes

**Comment by SyncIt21** [comment]
> The same can be said for the release of air when normal pipes are unwrenched. It is quite unpleasant to lose a part of some gas at each interaction with components and this can be used against others.

Yes we know it is unpleasant, good. Not all changes have to be for the benifit of the player, having a large amount of gas that sends players flying or an small amount of toxic gas that slowly spreads out and kills everyone when you unwrench a pipe is not only realistic but also funny.

It also punishes players in a way that teaches them not to make the same dumb mistakes again and forces them to make better atmos setups such as pumps to properly drain out the gas from the pipes before unwrenching them

Not to mention the damages it causes encourages interactions among other departments to come help clean the mess.

1)A small/large amount of zakur(Poisonous) gas released when you unwrench a pipe as it poisons players causes them to run to medbay for a cure thus giving medical something to do

2)A large/small amount of plasma(Flammable) gas released when you unwrench a pipe as it causes fire causes both environmental and people damage therefore encouraging engineering to come down and fix the damages and medbay to heal players of their burn damage

All these cases encourages inter departmental interactions which you are removing and making the game less fun for other departments by making atmos safe.

Remove the code for merging gases with neighbouring pipes and your PR should be good to go else people are going to debate about it till your PR goes stale and auto closes

**Comment by SyncIt21** [comment]
> Zauker decomposes instantly when it comes into contact with nitrogen, so this scenario is basically impossible.

This is just an example, sure I'll admit I'm not accurate on this but the point has been made

> I've never seen engineering actually fix the damage caused by a plasma fire. It causes almost as much damage as a breach, ask yourself how often you see engineers actually fix those.

This is just your personal observation, and doesn't sum up every players action in these circumstances. It provides an opportunity wether people take it or not is irrelevant and a skill issue not an coding one

**Comment by SyncIt21** [comment]
> > Remove the code for merging gases with neighbouring pipes
> 
> Should I also revert [default rotations](https://github.com/tgstation/tgstation/blob/ebd3db3e3197bef8373dd144f9513dd4e0357611/code/modules/atmospherics/machinery/components/components_base.dm#L269)?
> 
> [commit](https://github.com/tgstation/tgstation/commit/1b10930844f2d9e0ef906f25755b2abed86851bb)

Don't revert the entire commit just remove the proc relocate_airs() and all its uses. Then update your branch with master so we can re review ot

**Comment by FlufflesTheDog** [comment]
> But if you merge the air with neighbouring connected components (for e.g. a pipe that is connected in all 4 directions) then you now have to deal with emptying gases from 4 pipes and you now have at worst 4 times the work to do (if all the pipes are interconnected sure then the work is the same but if not then its a lot of extra tedium) so it causes much more problems than it solves.

This doesn't make any sense. If unwrenching a device would have moved air into a pipe, that pipe would already have had the exact gas mixture. The original change wouldn't ever make cleaning out **pipes** any more work than it already is, because it only moves gas to places its already directly connected to. It also didn't apply to pipe unwrenching like in your example.

> You are not looking at the Bigger picture i.e. a chain reaction of events that could occur from minor gas leaks.

Most of this stuff will only be relevant in atmospherics, the single department most capable of dealing with the problems that arise from it. Sure, people will be caught off guard, make mistakes or not know how to deal with certain things, but that already happens plenty with the many other destructive mechanics that atmos techs have to deal with, where clicking a button in the wrong order will dump the entirety of your project into the atmosphere. Ordnance is the one other place that touches atmos, and they're not too far behind on the tools they have for dealing with mistakes

> All in all if people want to make creating atmos setups more efficient/faster then they need to innovate by creating new types of components/setups.

People do things the way they do not just because it's quick and easy, but because making an actual robust setup is generally *less* efficient, and slower in terms of reaction speed, than the other methods. Did you know that just about every component, pump/vent/scrubber has a 200L volume, on each node it connects to the network? That's about 40% more than exists in a pipe, at 140L. And the entirety of that volume acts as dead weight when doing reactions, or heating and cooling, or emptying a pipe network.

This volume contributes a lot to making a big imbalance in the consequences of moving said devices. Unwrenching a pump (400L total) in the middle of a delam, for example, could release nearly three times as much gas into the air as unwrenching a pipe. Beyond making an already extremely hostile-to-beginners department more punishing of mistakes, I don't think such behavior makes much sense.

**Comment by SyncIt21** [comment]
> This doesn't make any sense. If unwrenching a device would have moved air into a pipe, that pipe would already have had the exact gas mixture

Pipes true I didn't meant to use that as an example my bad. Connector ports not so much which brings us back to Mtandi point

> Most of this stuff will only be relevant in atmospherics, 

And that's all that matters. If tiles are pried out and the pipes underneath are exposed in other departments then you have a greater impact.
Not to mention atmos is the only department that is connected to the whole station so if something like the distro or waste distribution pipes are destroyed by these mishapes then then everyone is effected.

> People do things the way they do not just because it's quick and easy, but because making an actual robust setup is generally less efficient, and slower in terms of reaction speed, than the other methods. Did you know that just about every component, pump/vent/scrubber has a 200L volume, on each node it connects to the network? That's about 40% more than exists in a pipe, at 140L. And the entirety of that volume acts as dead weight when doing reactions, or heating and cooling, or emptying a pipe network.

All these words such as "less efficient" "slower" can be chalked up to 3 words "just your opinion". Maybe you need to reinvent your existing setups and get better

If there is a problem with scrubbers having more volume than it should then create a seperate PR specifically talking about it's problem and how to fix that, I agree with your numbers on that but others can disagree and so that's a seperate issue not related to this PR

> Unwrenching a pump (400L total) in the middle of a delam, for example, could release nearly three times as much gas into the air as unwrenching a pipe

The  don't unwrench the pipe in the middle of the delam LOL!. people who do have a death wish and should be gitbanned.
A sound minded technician would have figured out a way to use volume pumps to drain the hot gases out quickly and then stabalize the HFR rather than unwrenching super pressurized pipes, skill issue  not coding.

Overall all your points about scrubbers having 200L and pipes having 140L will release a lot of gas true, so what's the solution?
Use pumps to drain the gases out from them and make them safe. Maybe a maintainer might not want components like scrubbers or pumps to release gases but rather simply store them until they are reconnected I'm cool with that but when it comes to pipes there are no excuses.

**Comment by LemonInTheDark** [comment]
Unwrenching components not dumping gas everywhere is both a quality of life thing, and also sort of an abstraction.
Things like pumps, connectors, and filters only "hold" gas in any given moment because they borrow it from their connected pipenets. 

That's why the volume doesn't really make sense, because volume on components is sequentially a way they "request" gas from their connected pipelines.

My gut is unwrenching pipes should remain as it is, and components should unwrench and dump their gas into neighboring pipes depending on the pipenet they face, or into the air if none exists. I realize this is sort of a vibes loss, but it allows for more flexibility for players. Atmos is already hard to design setups for, I don't wanna make it worse, and I don't want to invalidate things like thermomachines being movable. 

I don't consider the current state of things to be a massive issue, of grief or of like vibes, but it could always be improved a bit.

Sorry for bein MIA for so long, been kinda busy.

**Comment by mogeoko** [comment]
Done some not-so-great things. I apologize and will close the PR for the time being. I'll rework it and upload it later.

**Comment by LemonInTheDark** [comment]
no worries, this happens a lot more often then you'd think. lemme know if you need any help with it yeah?

**Pull request closed without merging**